src/libsync/discoveryphase.cpp

```diff
@@ -29,36 +52,59 @@ bool DiscoveryJob::isInSelectiveSyncBlackList(const QString& path) const
         return false;
     }
 
-    // If one of the item in the black list is a prefix of the path, it means this path need not to
-    // be synced.
-    //
-    // We know the list is sorted (for it is done in DiscoveryJob::start)
-    // So we can do a binary search. If the path is a prefix if another item or right after in the lexical order.
+    // Block if it is in the black list
+    return findPathInList(_selectiveSyncBlackList, path);
 
-    QString pathSlash = path + QLatin1Char('/');
+}
 
-    auto it = std::lower_bound(_selectiveSyncBlackList.begin(), _selectiveSyncBlackList.end(), pathSlash);
+int DiscoveryJob::isInSelectiveSyncBlackListCallback(void *data, const char *path)
+{
+    return static_cast<DiscoveryJob*>(data)->isInSelectiveSyncBlackList(QString::fromUtf8(path));
+}
 
-    if (it != _selectiveSyncBlackList.end() && *it == pathSlash) {
-        return true;
+bool DiscoveryJob::checkSelectiveSyncNewShare(const QString& path)
+{
+    // If this path or the parent is in the white list, then we do not block this file
+    if (findPathInList(_selectiveSyncWhiteList, path)) {
+        return false;
     }
 
-	if (it == _selectiveSyncBlackList.begin()) {
+    if (_newSharedFolderSizeLimit < 0) {
+        // no limit, everything is allowed;
         return false;
     }
-    --it;
-    Q_ASSERT(it->endsWith(QLatin1Char('/'))); // Folder::setSelectiveSyncBlackList makes sure of that
-    if (pathSlash.startsWith(*it)) {
+
+    // Go in the main thread to do a PROPFIND to know the size of this directory
+    qint64 result = -1;
+
+    {
+        QMutexLocker locker(&_vioMutex);
+        emit doGetSizeSignal(path, &result);
```

Comment:
Should we guard against going here for each discovery? In particular, should we do this if the folder already is in the third 'undecided' list?